### PR TITLE
knightos-scas: fix cross compilation and enable documentation

### PR DESCRIPTION
**Pull request by siraben**
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 131826` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>knightos-z80e</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>knightos-scas</li>
  </ul>
</details>

**Comment by siraben** [comment]
Not the cleanest fix, but that's because

- the make-based build fails on darwin (except when cross compiling)
- I can only fix cross compilation when using the make-based build

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 131826` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>knightos-z80e</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>knightos-scas</li>
  </ul>
</details>

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 131826` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>knightos-scas</li>
    <li>knightos-z80e</li>
  </ul>
</details>

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 131826` at 2be222ca run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/4ad7b9edc539d5688e8799d72ec17fdb">knightos-z80e</a></li>
  </ul>
</details>
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>knightos-scas</li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Consider adding a comment explaining the purpose of this patch on the line preceeding.
Near pkgs/development/tools/knightos/scas/default.nix:23:15:
```
   |
23 |   patches = [ ./cross.patch ];
   |               ^
```
</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

**Comment by Mindavi** [comment]
LGTM in general, but indeed knightos-z80e is broken on x86. I'm looking into it, but I'll have to build some stuff first due to the zstd brokenness on master.

This PR also enables documentation, might be good to note in the commit message too.

**Comment by Mindavi** [comment]
Been looking a bit more into this, why has the building switched from cmake to make? Is that to support cross-compilation?

Anyhow, that switch has broken / changed the output folder where the files are stored:

[diffoscope output](https://github.com/NixOS/nixpkgs/files/6895735/output.html.txt)

**Comment by Mindavi** [comment]
What about this instead?

```patch
diff --git a/pkgs/development/tools/knightos/scas/default.nix b/pkgs/development/tools/knightos/scas/default.nix
index 255ecf798d8..c62cd3cda73 100644
--- a/pkgs/development/tools/knightos/scas/default.nix
+++ b/pkgs/development/tools/knightos/scas/default.nix
@@ -1,4 +1,4 @@
-{ fetchFromGitHub, lib, stdenv, cmake }:
+{ fetchFromGitHub, lib, stdenv, cmake, buildPackages }:
 
 stdenv.mkDerivation rec {
   pname = "scas";
@@ -13,8 +13,16 @@ stdenv.mkDerivation rec {
 
   cmakeFlags = [ "-DSCAS_LIBRARY=1" ];
 
+  postPatch = ''
+    substituteInPlace CMakeLists.txt \
+      --replace "TARGETS scas scdump scwrap" "TARGETS scas scdump scwrap generate_tables"
+  '';
+
   strictDeps = true;
 
+  depsBuildBuild = lib.optionals (stdenv.buildPlatform != stdenv.hostPlatform) [
+    buildPackages.knightos-scas
+  ];
   nativeBuildInputs = [ cmake ];
 
   meta = with lib; {
```

Might need some darwin fixes, but this does the following:

- Build tools for native
- Use native tools for cross-compilation

It would be nicer to move the generate_tables to a separate output, but this is already a good start, I think.

**Comment by siraben** [comment]
Thanks! I adapted your patch to ensure the documentation also installs. Cross and native builds work on darwin and linux now.

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 131826` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>knightos-scas</li>
    <li>knightos-z80e</li>
  </ul>
</details>

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 131826` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>knightos-scas</li>
    <li>knightos-z80e</li>
  </ul>
</details>

File: pkgs/development/tools/knightos/scas/default.nix [hunk 1]
```diff
@@ -12,10 +16,18 @@ stdenv.mkDerivation rec {
...
-
+  postPatch = ''
+    substituteInPlace CMakeLists.txt \
+      --replace "TARGETS scas scdump scwrap" "TARGETS scas scdump scwrap generate_tables"
+  '';
   strictDeps = true;
 
-  nativeBuildInputs = [ cmake ];
+  depsBuildBuild = lib.optionals isCrossCompiling [ buildPackages.knightos-scas ];
+  nativeBuildInputs = [ asciidoc libxslt.bin cmake ];
```
Comment (SuperSandro2000) [inline pkgs/development/tools/knightos/scas/default.nix:26]: Do we need libxslt and asciidoc all the time or just for cross?

Comment (siraben) [reply pkgs/development/tools/knightos/scas/default.nix:26]: It's needed for documentation

**Comment by siraben** [comment]
@Mindavi can we make it so that generate_tables is not installed? It's supposed to be a build-time only tool.

**Review by SuperSandro2000 — APPROVED** [review]
diff LGTM

**Comment by Mindavi** [comment]
We cannot do that when building it this way. We could move it to a different output though, like 'dev'. Since cmake doesn't know the concept of building for the 'build' system, we're now using the built tool from the 'native' built version for cross-compilation.

With make you could indeed build it during the cross-compilation, not depending on the 'native' version first. But in that case some more fixes are probably needed to the makefile.

With cmake I don't think there are any great solutions, it just doesn't have a concept for this otherwise. Meson and make do support it though.

**Comment by siraben** [comment]
Ok, AFAIK the only issues with the Makefile are when compiling on darwin. I'll paste some logs.

**Comment by Mindavi** [comment]
Still builds fine when cross-compiling from x86_64 to aarch64. Tried both from the direct commit and from a rebase to master.

**Pull request merged**